ml-kem/src/compress.rs

```diff
@@ -85,36 +91,32 @@ pub(crate) mod test {
     use super::*;
     use hybrid_array::typenum::{U1, U10, U11, U12, U4, U5, U6};
 
-    // Verify that the integer compression routine produces the same results as rounding with
-    // floats.
-    fn compression_known_answer_test<D: CompressionFactor>() {
-        let fq: f64 = FieldElement::Q as f64;
-        let f2d: f64 = 2.0_f64.powi(D::I32);
-
+    // Verify against inequality 4.7
+    fn compression_decompression_inequality<D: CompressionFactor>() {
         for x in 0..FieldElement::Q {
-            let fx = x as f64;
-            let mut x = FieldElement(x);
+            let mut y = FieldElement(x);
 
-            // Verify equivalence of compression
-            x.compress::<D>();
-            let fcx = ((f2d / fq * fx).round() as Integer) % (1 << D::USIZE);
-            assert_eq!(x.0, fcx);
+            y.compress::<D>();
+            y.decompress::<D>();
 
-            // Verify equivalence of decompression
-            x.decompress::<D>();
-            let fdx = (fq / f2d * (fcx as f64)).round() as Integer;
-            assert_eq!(x.0, fdx);
+            let lhs = (i32::from(y.0) - i32::from(x)) % (i32::from(FieldElement::Q));
+            let rhs = ((FieldElement::Q32 + (1 << (D::U32 + 1)) - 1) / (1 << (D::U32 + 1))) as i32;
+
+            assert!(
+                lhs <= rhs,
```

Comment:
```suggestion
                lhs.abs() <= rhs,
```

Otherwise, the error can be arbitrarily large in the negative direction.  The spec allows this, but I think that's a bug in the spec.

Reply:
I don't think taking the absolute value works here as we are working over a field. In this case, `(x % q).abs() != x mod q` for negative x.

On further thinking, I believe we are instead looking for a symmetric modulo and propose:

```
            let mut error = (i32::from(y.0) - i32::from(x)) % half_q;
            if error < - half_q {
                error += half_q;
            }

            assert!(
                error <= error_threshold,
                "Inequality failed for x = {x}: error = {error}, error_threshold = {error_threshold}, D = {:?}",
                D::USIZE
            );
```
(where HALF_Q is defined outside the loop)

thoughts?

Reply:
We're not actually working in a finite field here, though.  The boolean `error < -half_q` is always false, given how the mod operation works.

How about this approach: It should always be the case that `x >= Decompress(Compress(x))`, since the effect of `Decompress(Compress(x))` is basically to zero out the low-order bits of `x`.  So you could do something like:

``` rust
assert!(x >= y.0);
assert!((x - y.0) % Self::Q < error_threshold);
```

It seems like we ought to be able to do an exhaustive known-answer test here, though.  Maybe you could try checking that `y.0 = x & ~((1 << D) - 1)`.